### DISCUSSION TITLE
feat: set screen options for each model

### DISCUSSION THREAD

**Discussion by grabbou** [rise-tools/rise-tools#119] [Design Doc]
Previously, this issue was related to migrating from Expo Router to React Navigation:

> This is an ongoing umbrella issues. Features or bugs will be added on an ongoing basis to track importance:
>
> - Not being able to choose screen presentation - there is no filesystem and/or programatic API to do so
> - Not being able to define all screens ahead of time. We have list of models. We could get list of all the screens and declare them, together with initial options. That way we could push to a screen that hasn't loaded yet w/o a flickering title (right now, we will put default title and then, update the title again once it renders with <Stack.Title />)

After discussing this with @ericvicenti, turns out there's different challenge here and is the lack of statically accessible, available ahead of time, information about each route in the app together with its options.

If there was such a map, essentially the following would be possible regardless of the routing mechanism under the hood. If we can find a solution using current navigation library, that's great!

**Comment by ericvicenti** [comment]
screen presentation, you want to use modals? I guess you looked into this: https://docs.expo.dev/router/advanced/modals/

for defining screens ahead of time I think it can be done with expo router. the tricky part is that we don't have the metadata like title handy at the time when we push to that route. if we did, there are ways

**Comment by grabbou** [comment]
Yeah, it kinda sucks but I think it could be possible with Expo Router to do the following:

```tsx
import { Stack } from 'expo-router';
export default function Layout() {
  return (
    <Stack>
      {screens.map(screen => (
        <Stack.Screen
          name={screen.name}
          options={screen.options}
        />
      )}
    </Stack>
  );
}
```

However, [as per their documentation](https://docs.expo.dev/router/advanced/modals/), it appears all screens need to have a corresponding file on the filesystem, which is not the case with React Navigation. So, unless I am wrong, we would have to move away from Expo Router or find an API to register a screen programatically (there's one for sure, question whether it's exported).

To make this work, we need to solve:

> for defining screens ahead of time I think it can be done with expo router. the tricky part is that we don't have the metadata like title handy at the time when we push to that route. if we did, there are ways

Do you have any thoughts? @ericvicenti

**Comment by grabbou** [comment]
I have an idea (maybe stupid, but worth checking if possible): In your default model, e.g. "inventory", define "Stack" and "Stack Screen" however you want.

```tsx
import { Stack } from '@rise-tools/kit-navigation';
export const models = {
   inventory: Inventory,
   inventoryHome: inventoryHome,
   inventoryItem: InventoryItem,
}

export function Inventory() {
  return (
    <Stack>
      <Stack.Screen
          name="inventoryHome"
          options={{ title: "Inventory" }}
      />
     <Stack.Screen
          name="inventoryItem"
          options={{ title: "Inventory Item" }}
      />
    </Stack>
  );
}
```

I think if this worked, it could potentially unlock unlimited customisation and defining different nested stack navigators, tabs etc. as you please.

> Note: This assumes Expo Router does not require file on the filesystem for each route or group of routes and works similar to React Navigation.
